src/cache/azure.rs

```diff
@@ -25,16 +25,20 @@ use crate::errors::*;
 pub struct AzureBlobCache {
     container: Arc<BlobContainer>,
     credentials: AzureCredentials,
+    key_prefix: String,
 }
 
 impl AzureBlobCache {
-    pub fn new() -> Result<AzureBlobCache> {
-        let credentials = match EnvironmentProvider.provide_credentials() {
-            Ok(creds) => creds,
-            Err(err) => bail!(
-                "Could not find Azure credentials in the environment: {}",
-                err
-            ),
+    pub fn new(credentials: Option<AzureCredentials>, key_prefix: &str) -> Result<AzureBlobCache> {
```

Comment:
Good call on making `credentials` be provided by the consumer to make your tests possible to write.

It's a little funky seeing the `None` get provided into `AzureBlobCache` in `storage_from_config()` down below: perhaps it's worth trying one of the following two options:
1. Make `credentials` mandatory, and move the default `EnvironmentProvider.provide_credentials()` up to the caller
2. (My preferred option): Make two separate functions here: `pub fn new(credentials: AzureCredentials, key_prefix: &str)` and `pub fn new_with_environment_credentials(key_prefix: &str)`.

Let me know what you think :)

Reply:
I'm fine with either, so I'll implement the second.

Reply:
~I'm fine with either, so I'll implement the second.~
On second thought, I think the first one is better going forward. The problem with the second one is that, as more options get introduced for AzureStorage, you need to duplicate all parameters twice.